src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapSessionOptions.Linux.cs

```diff
@@ -52,6 +52,13 @@ public ReferralChasingOptions ReferralChasing
             }
         }
 
+        public void StartTransportLayerSecurity(DirectoryControlCollection controls)
+        {
+            if (_connection._disposed) throw new ObjectDisposedException(GetType().Name);
+            _connection._startTls = true;
+            _connection._startTlsControls = controls;
```

Comment:
There's a mutability concern with the collection here.

In the Windows version the collection is read during this method and turned into native values, no modifications to the collection matter (including any modifications to items in the collection (assuming they're mutable)).

In the Linux version, since just a soft reference to the collection is kept to the collection, if the collection was modified between the call to StartTLS and Bind that would change how the StartTLS functioned.

There's also some state complexity with this approach.

```C#
conn.SessionOptions.StartTransportLayerSecurity(opts1);
conn.SessionOptions.StartTransportLayerSecurity(opts2);
```

In Windows I believe this throws, because StartTLS can't be called again after it finishes (maybe unless there was a StopTLS, but that's it's own can of worms).  With this change the Linux behavior would be that the opts2 options are what get used, not opts1.

Reply:
> In Windows I believe this throws, because StartTLS can't be called again after it finishes (maybe unless there was a StopTLS, but that's it's own can of worms).

Indeed, it does. Cf. #49726.
> In the Linux version, since just a soft reference to the collection is kept to the collection, if the collection was modified between the call to StartTLS and Bind that would change how the StartTLS functioned.
> There's also some state complexity with this approach. ... With this change the Linux behavior would be that the opts2 options are what get used, not opts1.

Oh, I see. If we moved the native copy/marshaling steps that are currently in LdapConnection.Linux.cs, we could have the same semantics (even though they wouldn't be used until later), right?

To match the Windows behavior, could something like this work?
```csharp
    public void StartTransportLayerSecurity(DirectoryControls controls)
    {
        lock(_connection._startTlsLock)
        {
            if (_connection._startTls) throw new TlsOperationException() : true;
            _connection._startTls = true;
        }
   
        _connection._startTlsServerControlsPtr = CreateManagedControls(BuildControlArray(controls, true));
        _connection._startTlsClientControlsPtr = CreateManagedControls(BuildControlArray(controls, false));
    }
    // ...
    IntPtr CreateManagedControls(DirectoryControl[] controls)
    {
        // ...
    }

    void FreeManagedControls(IntPtr controls)
    {
        // ...
    }

```

`FreeManagedControls()` would be called after `LdapPal.StartTls()` or on Dispose of the `LdapConnection`. I think this would ensure that the operation can only get called once or else throw an exception, and since the `controls` argument is being copied, the value wouldn't be subject to mutability.

I'm digger deeper into the OpenLDAP code, and I think that the `ldap_install_tls` OpenLDAP function may give us another option. I'll have to investigate that more.


If this deferral method doesn't work (which it seems like it may not based on your other comments about exceptions), then there's another option, as mentioned in the linked issue:

> During `Init()`, initialize the LDAP handle with the `ldap://` scheme (or `cldap://` since `LdapDirectoryIdentifier.Connectionless` is known at that time). Then during `Connect()`, check if `SecureSocketsLayer` is set, and then rebuild the URI.
> - This means we'd have to build the URI list twice if SecureSocketsLayer is set.
> - If there are no other pre-bind operations that can take place using S.DS.P, then this might be an OK option.

I'll also add that since it's [almost?] always a mistake to configure a connection to use both opportunistic _and_ explicit TLS, these two methods shouldn't conflict.